enterprise/server/remote_execution/runner/runner_linux.go

```diff
@@ -0,0 +1,124 @@
+	}
+	if dockerProvider != nil {
+		providers[platform.DockerContainerType] = dockerProvider
+	}
+	podmanProvider, err := podman.NewProvider(p.env, p.imageCacheAuth, *rootDirectory)
+	if err != nil {
+		return status.FailedPreconditionErrorf("Failed to initialize podman container provider: %s", err)
+	}
+	if podmanProvider != nil {
+		providers[platform.PodmanContainerType] = podmanProvider
+	}
+	providers[platform.FirecrackerContainerType] = firecracker.NewProvider(p.env, p.imageCacheAuth, *rootDirectory)
+	providers[platform.BareContainerType] = &bare.Provider{}
+
+	p.containerProviders = providers
+
+	return nil
+}
+
+func (r *commandRunner) startVFS() error {
+	var fs *vfs.VFS
+	var vfsServer *vfs_server.Server
+	enableVFS := r.PlatformProperties.EnableVFS
+	// Firecracker requires mounting the FS inside the guest VM so we can't just swap out the directory in the runner.
+	if enableVFS && platform.ContainerType(r.PlatformProperties.WorkloadIsolationType) != platform.FirecrackerContainerType {
+		vfsDir := r.Workspace.Path() + "_vfs"
+		if err := os.Mkdir(vfsDir, 0755); err != nil {
+			return status.UnavailableErrorf("could not create FUSE FS dir: %s", err)
+		}
+
+		vfsServer = vfs_server.New(r.p.env, r.Workspace.Path())
+		unixSocket := filepath.Join(r.Workspace.Path(), "vfs.sock")
+
+		lis, err := net.Listen("unix", unixSocket)
+		if err != nil {
+			return err
+		}
+		if err := vfsServer.Start(lis); err != nil {
+			return err
+		}
+
+		conn, err := grpc.Dial("unix://"+unixSocket, grpc.WithInsecure())
+		if err != nil {
+			return err
+		}
+		vfsClient := vfspb.NewFileSystemClient(conn)
+		fs = vfs.New(vfsClient, vfsDir, &vfs.Options{})
+		if err := fs.Mount(); err != nil {
+			return status.UnavailableErrorf("unable to mount VFS at %q: %s", vfsDir, err)
+		}
+	}
+
+	r.VFS = fs
+	r.VFSServer = vfsServer
+	return nil
+}
+
+func (r *commandRunner) prepareVFS(ctx context.Context, layout *container.FileSystemLayout) error {
+	if r.PlatformProperties.EnableVFS {
+		// Unlike other "container" implementations, for Firecracker VFS is mounted inside the guest VM so we need to
+		// pass the layout information to the implementation.
+		if fc, ok := r.Container.Delegate.(*firecracker.FirecrackerContainer); ok {
+			fc.SetTaskFileSystemLayout(layout)
+		}
+	}
+
+	if r.VFSServer != nil {
+		p, err := vfs_server.NewCASLazyFileProvider(r.env, ctx, layout.RemoteInstanceName, layout.DigestFunction, layout.Inputs)
+		if err != nil {
+			return err
+		}
+		if err := r.VFSServer.Prepare(p); err != nil {
+			return err
+		}
+	}
+	if r.VFS != nil {
+		if err := r.VFS.PrepareForTask(ctx, r.task.GetExecutionId()); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *commandRunner) removeVFS() []error {
+	errs := []error{}
+	if r.VFS != nil {
+		if err := r.VFS.Unmount(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if r.VFSServer != nil {
+		r.VFSServer.Stop()
+	}
+	return errs
+}
```

Comment:
nit: `[]error` feels a bit unconventional as a return value in go. If you want to `log.Error()` and return the last non-nil error (e.g. `if err != nil { lastErr = err } ... return lastErr`), I think that works.

Reply:
This is mostly because of the current way error handling works in `*commandRunner.Remove()`, which collect a slice of errors and then returns an `errSlice`, which are multiple errors concat into 1.

So I fixed `removeVFS` to return errSlice as well, mostly so that we could preserve the data of all errors in the slice, which could be valuable when troubleshooting.

Reply:
nvm I see the problem 🙈 
will fix!